src/systems/rtti.rs

```diff
@@ -0,0 +1,312 @@
+use crate::raw::root::RED4ext as red;
+use crate::{Array, Bitfield, CName, Class, Enum, Function, Type};
+
+#[repr(transparent)]
+pub struct CRTTISystem(*mut red::CRTTISystem);
+
+impl CRTTISystem {
+    pub fn get() -> Self {
+        unsafe { Self(red::CRTTISystem_Get()) }
+    }
+
+    #[inline]
+    pub fn get_class(&self, name: CName) -> Option<&Class> {
+        let class = unsafe { (self.vft().base.get_class)(&(*self.0)._base, name.0) };
```

Comment:
```suggestion
        let class = unsafe { (self.vft().base.get_class)(self, name) };
```
I'd probably
- define the vft to use our wrappers since they're `repr(transparent)` anyway and will simplify usage for us
- `&(*self.0)._base` doesn't actually do anything since base is the first field, so it can be written as `&self.0`

Reply:
so I made the changes in bda5bf1 and if I understand you correctly, we can indeed do:
```rs
#[inline]
pub fn get_class(&mut self, name: CName) -> Option<&Class> {
    let ty = unsafe { (self.vft().get_class)(self, name) };
    unsafe { ty.cast::<Class>().as_ref() }
}
```
where indeed by taking `&mut` we can just use `self` instead of `self as *const _ as *mut _`.
I love its simplicity, except that on usage it loses in ergonomy, e.g. :
![Screenshot 2024-06-26 163838](https://github.com/jac3km4/red4rs/assets/21016014/129d9d59-88de-4062-920c-990068b58f82)
or maybe there's an alternative that I overlooked?

Reply:
but at the same time it's perfectly fine to do:
```rs
fn inspect_derived_classes(name: &str) {
    let env = TestRed4rs::env();
    if let Some(parent) = RttiSystem::get().get_class(CName::new(name)) {
        let classes = RttiSystem::get().get_derived_classes(parent);
        log::info!(
            env,
            "derived classes for {name}: {}",
            classes
                .iter()
                .map(|x| x.name().as_str())
                .collect::<Vec<_>>()
                .join(", ")
        );
    } else {
        log::info!(env, "couldn't find derived class for {name}");
    }
}
```
which is maybe what you have in mind?

Reply:
if this is the case, we could also keep `get()` private and make the other methods static, e.g.:
```rs
impl RttiSystem {
    fn get() -> Self {
        Self(unsafe { std::ptr::read(red::CRTTISystem_Get()) }._base)
    }

    #[inline]
    pub fn get_class(&mut self, name: CName) -> Option<&Class> {
        unsafe { (Self::get().vft().get_class)(self, name).cast::<Class>().as_ref() }
    }
    // ...
}

Reply:
I see, it might make sense to have the getters that are known not to mutate anything accept & instead of &mut, you can either cast it using `as` or change the vtable to use *const for the `this` parameter

Reply:
yes the issue at the moment is that since the latest changes the game crashes with a:
```
Error reason: Unhandled exception
Expression: EXCEPTION_ACCESS_VIOLATION (0xC0000005)
Message: The thread attempted to read inaccessible data at 0x18.
File: <Unknown>(0)
```

Reply:
I'm pretty sure that's because you're reading the value behind the pointer for RttiSystem, see my other comment above

Reply:
Ok I isolated the commit where the issue appeared: everything works fine up to 3ac4353.